File: artifacts/heman/refactor/aabb3.c
#include <aabb3.c>
#include <aabb3.h>
#include <assert.h>
#include <math.h>
#include <utility.h>
#include <vec3.h>

extern float kmSQR(float s);
extern float kmDegreesToRadians(float degrees);
extern float kmRadiansToDegrees(float radians);
extern float kmMin(float lhs, float rhs);
extern float kmMax(float lhs, float rhs);
extern unsigned char kmAlmostEqual(float lhs, float rhs);
extern float kmClamp(float x, float min, float max);
extern float kmLerp(float x, float y, float factor);
struct kmMat4;
struct kmMat3;
struct kmPlane;
typedef struct kmVec3
{
  float x;
  float y;
  float z;
} kmVec3;
kmVec3 *kmVec3Fill(kmVec3 *pOut, float x, float y, float z);
float kmVec3Length(const kmVec3 *pIn);
float kmVec3LengthSq(const kmVec3 *pIn);
kmVec3 *kmVec3Lerp(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2, float t);
kmVec3 *kmVec3Normalize(kmVec3 *pOut, const kmVec3 *pIn);
kmVec3 *kmVec3Cross(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2);
float kmVec3Dot(const kmVec3 *pV1, const kmVec3 *pV2);
kmVec3 *kmVec3Add(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2);
kmVec3 *kmVec3Subtract(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2);
kmVec3 *kmVec3Mul(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2);
kmVec3 *kmVec3Div(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2);
kmVec3 *kmVec3MultiplyMat3(kmVec3 *pOut, const kmVec3 *pV, const struct kmMat3 *pM);
kmVec3 *kmVec3MultiplyMat4(kmVec3 *pOut, const kmVec3 *pV, const struct kmMat4 *pM);
kmVec3 *kmVec3Transform(kmVec3 *pOut, const kmVec3 *pV1, const struct kmMat4 *pM);
kmVec3 *kmVec3TransformNormal(kmVec3 *pOut, const kmVec3 *pV, const struct kmMat4 *pM);
kmVec3 *kmVec3TransformCoord(kmVec3 *pOut, const kmVec3 *pV, const struct kmMat4 *pM);
kmVec3 *kmVec3Scale(kmVec3 *pOut, const kmVec3 *pIn, const float s);
int kmVec3AreEqual(const kmVec3 *p1, const kmVec3 *p2);
kmVec3 *kmVec3InverseTransform(kmVec3 *pOut, const kmVec3 *pV, const struct kmMat4 *pM);
kmVec3 *kmVec3InverseTransformNormal(kmVec3 *pOut, const kmVec3 *pVect, const struct kmMat4 *pM);
kmVec3 *kmVec3Assign(kmVec3 *pOut, const kmVec3 *pIn);
kmVec3 *kmVec3Zero(kmVec3 *pOut);
kmVec3 *kmVec3GetHorizontalAngle(kmVec3 *pOut, const kmVec3 *pIn);
kmVec3 *kmVec3RotationToDirection(kmVec3 *pOut, const kmVec3 *pIn, const kmVec3 *forwards);
kmVec3 *kmVec3ProjectOnToPlane(kmVec3 *pOut, const kmVec3 *point, const struct kmPlane *plane);
kmVec3 *kmVec3Reflect(kmVec3 *pOut, const kmVec3 *pIn, const kmVec3 *normal);
extern const kmVec3 KM_VEC3_NEG_Z;
extern const kmVec3 KM_VEC3_POS_Z;
extern const kmVec3 KM_VEC3_POS_Y;
extern const kmVec3 KM_VEC3_NEG_Y;
extern const kmVec3 KM_VEC3_NEG_X;
extern const kmVec3 KM_VEC3_POS_X;
extern const kmVec3 KM_VEC3_ZERO;
typedef struct kmAABB3
{
  kmVec3 min;
  kmVec3 max;
} kmAABB3;
kmAABB3 *kmAABB3Initialize(kmAABB3 *pBox, const kmVec3 *centre, const float width, const float height, const float depth);
int kmAABB3ContainsPoint(const kmAABB3 *pBox, const kmVec3 *pPoint);
kmAABB3 *kmAABB3Assign(kmAABB3 *pOut, const kmAABB3 *pIn);
kmAABB3 *kmAABB3Scale(kmAABB3 *pOut, const kmAABB3 *pIn, float s);
unsigned char kmAABB3IntersectsTriangle(kmAABB3 *box, const kmVec3 *p1, const kmVec3 *p2, const kmVec3 *p3);
unsigned char kmAABB3IntersectsAABB(const kmAABB3 *box, const kmAABB3 *other);
unsigned int kmAABB3ContainsAABB(const kmAABB3 *container, const kmAABB3 *to_check);
float kmAABB3DiameterX(const kmAABB3 *aabb);
float kmAABB3DiameterY(const kmAABB3 *aabb);
float kmAABB3DiameterZ(const kmAABB3 *aabb);
kmVec3 *kmAABB3Centre(const kmAABB3 *aabb, kmVec3 *pOut);
kmAABB3 *kmAABB3ExpandToContain(kmAABB3 *pOut, const kmAABB3 *pIn, const kmAABB3 *other);
kmAABB3 *kmAABB3Initialize(kmAABB3 *pBox, const kmVec3 *centre, const float width, const float height, const float depth)
{
  if (!pBox)
  {
    return 0;
  }
  kmVec3 origin;
  kmVec3 *point = (centre) ? ((kmVec3 *) centre) : (&origin);
  kmVec3Zero(&origin);
  pBox->min.x = point->x - (width / 2);
  pBox->min.y = point->y - (height / 2);
  pBox->min.z = point->z - (depth / 2);
  pBox->max.x = point->x + (width / 2);
  pBox->max.y = point->y + (height / 2);
  pBox->max.z = point->z + (depth / 2);
  return pBox;
}

int kmAABB3ContainsPoint(const kmAABB3 *pBox, const kmVec3 *pPoint)
{
  if ((((((pPoint->x >= pBox->min.x) && (pPoint->x <= pBox->max.x)) && (pPoint->y >= pBox->min.y)) && (pPoint->y <= pBox->max.y)) && (pPoint->z >= pBox->min.z)) && (pPoint->z <= pBox->max.z))
  {
    return 1;
  }
  return 0;
}

kmAABB3 *kmAABB3Assign(kmAABB3 *pOut, const kmAABB3 *pIn)
{
  kmVec3Assign(&pOut->min, &pIn->min);
  kmVec3Assign(&pOut->max, &pIn->max);
  return pOut;
}

kmAABB3 *kmAABB3Scale(kmAABB3 *pOut, const kmAABB3 *pIn, float s)
{
  assert(0 && "Not implemented");
  return pOut;
}

unsigned char kmAABB3IntersectsTriangle(kmAABB3 *box, const kmVec3 *p1, const kmVec3 *p2, const kmVec3 *p3)
{
  assert(0 && "Not implemented");
  return 1;
}

unsigned char kmAABB3IntersectsAABB(const kmAABB3 *box, const kmAABB3 *other)
{
  return kmAABB3ContainsAABB(box, other) != ((unsigned int) 0);
}

unsigned int kmAABB3ContainsAABB(const kmAABB3 *container, const kmAABB3 *to_check)
{
  kmVec3 corners[8];
  unsigned int result = (unsigned int) 2;
  unsigned char found = 0;
  kmVec3Fill(&corners[0], to_check->min.x, to_check->min.y, to_check->min.z);
  kmVec3Fill(&corners[1], to_check->max.x, to_check->min.y, to_check->min.z);
  kmVec3Fill(&corners[2], to_check->max.x, to_check->max.y, to_check->min.z);
  kmVec3Fill(&corners[3], to_check->min.x, to_check->max.y, to_check->min.z);
  kmVec3Fill(&corners[4], to_check->min.x, to_check->min.y, to_check->max.z);
  kmVec3Fill(&corners[5], to_check->max.x, to_check->min.y, to_check->max.z);
  kmVec3Fill(&corners[6], to_check->max.x, to_check->max.y, to_check->max.z);
  kmVec3Fill(&corners[7], to_check->min.x, to_check->max.y, to_check->max.z);
  for (unsigned char i = 0; i < 8; i += 1)
  {
    if (!kmAABB3ContainsPoint(container, &corners[i]))
    {
      result = (unsigned int) 1;
      if (found)
      {
        return result;
      }
    }
    else
    {
      found = 1;
    }
  }

  if (!found)
  {
    result = (unsigned int) 0;
  }
  return result;
}

float kmAABB3DiameterX(const kmAABB3 *aabb)
{
  return fabs(aabb->max.x - aabb->min.x);
}

float kmAABB3DiameterY(const kmAABB3 *aabb)
{
  return fabs(aabb->max.y - aabb->min.y);
}

float kmAABB3DiameterZ(const kmAABB3 *aabb)
{
  return fabs(aabb->max.z - aabb->min.z);
}

kmVec3 *kmAABB3Centre(const kmAABB3 *aabb, kmVec3 *pOut)
{
  kmVec3Add(pOut, &aabb->min, &aabb->max);
  kmVec3Scale(pOut, pOut, 0.5);
  return pOut;
}

kmAABB3 *kmAABB3ExpandToContain(kmAABB3 *pOut, const kmAABB3 *pIn, const kmAABB3 *other)
{
  kmAABB3 result;
  result.min.x = (pIn->min.x < other->min.x) ? (pIn->min.x) : (other->min.x);
  result.max.x = (pIn->max.x > other->max.x) ? (pIn->max.x) : (other->max.x);
  result.min.y = (pIn->min.y < other->min.y) ? (pIn->min.y) : (other->min.y);
  result.max.y = (pIn->max.y > other->max.y) ? (pIn->max.y) : (other->max.y);
  result.min.z = (pIn->min.z < other->min.z) ? (pIn->min.z) : (other->min.z);
  result.max.z = (pIn->max.z > other->max.z) ? (pIn->max.z) : (other->max.z);
  kmAABB3Assign(pOut, &result);
  return pOut;
}


extern float kmSQR(float s);
extern float kmDegreesToRadians(float degrees);
extern float kmRadiansToDegrees(float radians);
extern float kmMin(float lhs, float rhs);
extern float kmMax(float lhs, float rhs);
extern unsigned char kmAlmostEqual(float lhs, float rhs);
extern float kmClamp(float x, float min, float max);
extern float kmLerp(float x, float y, float factor);
struct kmMat4;
struct kmMat3;
struct kmPlane;
typedef struct kmVec3
{
  float x;
  float y;
  float z;
} kmVec3;
kmVec3 *kmVec3Fill(kmVec3 *pOut, float x, float y, float z);
float kmVec3Length(const kmVec3 *pIn);
float kmVec3LengthSq(const kmVec3 *pIn);
kmVec3 *kmVec3Lerp(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2, float t);
kmVec3 *kmVec3Normalize(kmVec3 *pOut, const kmVec3 *pIn);
kmVec3 *kmVec3Cross(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2);
float kmVec3Dot(const kmVec3 *pV1, const kmVec3 *pV2);
kmVec3 *kmVec3Add(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2);
kmVec3 *kmVec3Subtract(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2);
kmVec3 *kmVec3Mul(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2);
kmVec3 *kmVec3Div(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2);
kmVec3 *kmVec3MultiplyMat3(kmVec3 *pOut, const kmVec3 *pV, const struct kmMat3 *pM);
kmVec3 *kmVec3MultiplyMat4(kmVec3 *pOut, const kmVec3 *pV, const struct kmMat4 *pM);
kmVec3 *kmVec3Transform(kmVec3 *pOut, const kmVec3 *pV1, const struct kmMat4 *pM);
kmVec3 *kmVec3TransformNormal(kmVec3 *pOut, const kmVec3 *pV, const struct kmMat4 *pM);
kmVec3 *kmVec3TransformCoord(kmVec3 *pOut, const kmVec3 *pV, const struct kmMat4 *pM);
kmVec3 *kmVec3Scale(kmVec3 *pOut, const kmVec3 *pIn, const float s);
int kmVec3AreEqual(const kmVec3 *p1, const kmVec3 *p2);
kmVec3 *kmVec3InverseTransform(kmVec3 *pOut, const kmVec3 *pV, const struct kmMat4 *pM);
kmVec3 *kmVec3InverseTransformNormal(kmVec3 *pOut, const kmVec3 *pVect, const struct kmMat4 *pM);
kmVec3 *kmVec3Assign(kmVec3 *pOut, const kmVec3 *pIn);
kmVec3 *kmVec3Zero(kmVec3 *pOut);
kmVec3 *kmVec3GetHorizontalAngle(kmVec3 *pOut, const kmVec3 *pIn);
kmVec3 *kmVec3RotationToDirection(kmVec3 *pOut, const kmVec3 *pIn, const kmVec3 *forwards);
kmVec3 *kmVec3ProjectOnToPlane(kmVec3 *pOut, const kmVec3 *point, const struct kmPlane *plane);
kmVec3 *kmVec3Reflect(kmVec3 *pOut, const kmVec3 *pIn, const kmVec3 *normal);
extern const kmVec3 KM_VEC3_NEG_Z;
extern const kmVec3 KM_VEC3_POS_Z;
extern const kmVec3 KM_VEC3_POS_Y;
extern const kmVec3 KM_VEC3_NEG_Y;
extern const kmVec3 KM_VEC3_NEG_X;
extern const kmVec3 KM_VEC3_POS_X;
extern const kmVec3 KM_VEC3_ZERO;
typedef struct kmAABB3
{
  kmVec3 min;
  kmVec3 max;
} kmAABB3;
kmAABB3 *kmAABB3Initialize(kmAABB3 *pBox, const kmVec3 *centre, const float width, const float height, const float depth);
int kmAABB3ContainsPoint(const kmAABB3 *pBox, const kmVec3 *pPoint);
kmAABB3 *kmAABB3Assign(kmAABB3 *pOut, const kmAABB3 *pIn);
kmAABB3 *kmAABB3Scale(kmAABB3 *pOut, const kmAABB3 *pIn, float s);
unsigned char kmAABB3IntersectsTriangle(kmAABB3 *box, const kmVec3 *p1, const kmVec3 *p2, const kmVec3 *p3);
unsigned char kmAABB3IntersectsAABB(const kmAABB3 *box, const kmAABB3 *other);
unsigned int kmAABB3ContainsAABB(const kmAABB3 *container, const kmAABB3 *to_check);
float kmAABB3DiameterX(const kmAABB3 *aabb);
float kmAABB3DiameterY(const kmAABB3 *aabb);
float kmAABB3DiameterZ(const kmAABB3 *aabb);
kmVec3 *kmAABB3Centre(const kmAABB3 *aabb, kmVec3 *pOut);
kmAABB3 *kmAABB3ExpandToContain(kmAABB3 *pOut, const kmAABB3 *pIn, const kmAABB3 *other);
kmAABB3 *kmAABB3Initialize(kmAABB3 *pBox, const kmVec3 *centre, const float width, const float height, const float depth)
{
  if (!pBox)
  {
    return 0;
  }
  kmVec3 origin;
  kmVec3 *point = (centre) ? ((kmVec3 *) centre) : (&origin);
  kmVec3Zero(&origin);
  pBox->min.x = point->x - (width / 2);
  pBox->min.y = point->y - (height / 2);
  pBox->min.z = point->z - (depth / 2);
  pBox->max.x = point->x + (width / 2);
  pBox->max.y = point->y + (height / 2);
  pBox->max.z = point->z + (depth / 2);
  return pBox;
}

int kmAABB3ContainsPoint(const kmAABB3 *pBox, const kmVec3 *pPoint)
{
  if ((((((pPoint->x >= pBox->min.x) && (pPoint->x <= pBox->max.x)) && (pPoint->y >= pBox->min.y)) && (pPoint->y <= pBox->max.y)) && (pPoint->z >= pBox->min.z)) && (pPoint->z <= pBox->max.z))
  {
    return 1;
  }
  return 0;
}

kmAABB3 *kmAABB3Assign(kmAABB3 *pOut, const kmAABB3 *pIn)
{
  kmVec3Assign(&pOut->min, &pIn->min);
  kmVec3Assign(&pOut->max, &pIn->max);
  return pOut;
}

kmAABB3 *kmAABB3Scale(kmAABB3 *pOut, const kmAABB3 *pIn, float s)
{
  assert(0 && "Not implemented");
  return pOut;
}

unsigned char kmAABB3IntersectsTriangle(kmAABB3 *box, const kmVec3 *p1, const kmVec3 *p2, const kmVec3 *p3)
{
  assert(0 && "Not implemented");
  return 1;
}

unsigned char kmAABB3IntersectsAABB(const kmAABB3 *box, const kmAABB3 *other)
{
  return kmAABB3ContainsAABB(box, other) != ((unsigned int) 0);
}

unsigned int kmAABB3ContainsAABB(const kmAABB3 *container, const kmAABB3 *to_check)
{
  kmVec3 corners[8];
  unsigned int result = (unsigned int) 2;
  unsigned char found = 0;
  kmVec3Fill(&corners[0], to_check->min.x, to_check->min.y, to_check->min.z);
  kmVec3Fill(&corners[1], to_check->max.x, to_check->min.y, to_check->min.z);
  kmVec3Fill(&corners[2], to_check->max.x, to_check->max.y, to_check->min.z);
  kmVec3Fill(&corners[3], to_check->min.x, to_check->max.y, to_check->min.z);
  kmVec3Fill(&corners[4], to_check->min.x, to_check->min.y, to_check->max.z);
  kmVec3Fill(&corners[5], to_check->max.x, to_check->min.y, to_check->max.z);
  kmVec3Fill(&corners[6], to_check->max.x, to_check->max.y, to_check->max.z);
  kmVec3Fill(&corners[7], to_check->min.x, to_check->max.y, to_check->max.z);
  for (unsigned char i = 0; i < 8; i += 1)
  {
    if (!kmAABB3ContainsPoint(container, &corners[i]))
    {
      result = (unsigned int) 1;
      if (found)
      {
        return result;
      }
    }
    else
    {
      found = 1;
    }
  }

  if (!found)
  {
    result = (unsigned int) 0;
  }
  return result;
}

float kmAABB3DiameterX(const kmAABB3 *aabb)
{
  return fabs(aabb->max.x - aabb->min.x);
}

float kmAABB3DiameterY(const kmAABB3 *aabb)
{
  return fabs(aabb->max.y - aabb->min.y);
}

float kmAABB3DiameterZ(const kmAABB3 *aabb)
{
  return fabs(aabb->max.z - aabb->min.z);
}

kmVec3 *kmAABB3Centre(const kmAABB3 *aabb, kmVec3 *pOut)
{
  kmVec3Add(pOut, &aabb->min, &aabb->max);
  kmVec3Scale(pOut, pOut, 0.5);
  return pOut;
}

kmAABB3 *kmAABB3ExpandToContain(kmAABB3 *pOut, const kmAABB3 *pIn, const kmAABB3 *other)
{
  kmAABB3 result;
  result.min.x = (pIn->min.x < other->min.x) ? (pIn->min.x) : (other->min.x);
  result.max.x = (pIn->max.x > other->max.x) ? (pIn->max.x) : (other->max.x);
  result.min.y = (pIn->min.y < other->min.y) ? (pIn->min.y) : (other->min.y);
  result.max.y = (pIn->max.y > other->max.y) ? (pIn->max.y) : (other->max.y);
  result.min.z = (pIn->min.z < other->min.z) ? (pIn->min.z) : (other->min.z);
  result.max.z = (pIn->max.z > other->max.z) ? (pIn->max.z) : (other->max.z);
  kmAABB3Assign(pOut, &result);
  return pOut;
}

File: artifacts/heman/translate/src/aabb3.rs
use crate::*;
use std::assert;
use std::f32;

pub fn kmAABB3Scale<'a>(pOut: Option<&'a mut KmAabb3>, pIn: Option<&'a KmAabb3>, s: f32) -> Option<&'a mut KmAabb3> {
    // Use assert_eq! with a custom message to match C's assert behavior
    assert!(false, "Not implemented");
    
    // Return pOut as is (matching C behavior)
    pOut
}

pub fn kmAABB3DiameterX(aabb: Option<&KmAabb3>) -> f32 {
    // Check if the input is None (equivalent to NULL check in C)
    if aabb.is_none() {
        return 0.0; // Return 0.0 as a safe default for invalid input
    }

    let aabb = aabb.unwrap(); // Safe unwrap after checking
    f32::abs(aabb.max.x - aabb.min.x)
}

pub fn kmAABB3DiameterY(aabb: &KmAabb3) -> f32 {
    f32::abs(aabb.max.y - aabb.min.y)
}

pub fn kmAABB3DiameterZ(aabb: &KmAabb3) -> f32 {
    f32::abs(aabb.max.z - aabb.min.z)
}
pub fn kmAABB3ContainsPoint(pBox: &KmAabb3, pPoint: &KmVec3) -> bool {
    (pPoint.x >= pBox.min.x && pPoint.x <= pBox.max.x) &&
    (pPoint.y >= pBox.min.y && pPoint.y <= pBox.max.y) &&
    (pPoint.z >= pBox.min.z && pPoint.z <= pBox.max.z)
}
pub fn kmAABB3Assign<'a>(pOut: Option<&'a mut KmAabb3>, pIn: Option<&'a KmAabb3>) -> Option<&'a mut KmAabb3> {
    // Check for None (equivalent to NULL checks in C)
    if pOut.is_none() || pIn.is_none() {
        return None;
    }

    // Unwrap safely since we've checked for None
    let pOut = pOut.unwrap();
    let pIn = pIn.unwrap();

    // Assign min and max vectors
    kmVec3Assign(Some(&mut pOut.min), Some(&pIn.min));
    kmVec3Assign(Some(&mut pOut.max), Some(&pIn.max));

    Some(pOut)
}

pub fn km_aabb3_intersects_triangle(
    box_: &KmAabb3,
    p1: &KmVec3,
    p2: &KmVec3,
    p3: &KmVec3,
) -> u8 {
    assert!(false, "Not implemented");
    1
}
pub fn kmAABB3ContainsAABB(container: &KmAabb3, to_check: &KmAabb3) -> u32 {
    let mut corners: [KmVec3; 8] = [
        KmVec3 { x: 0.0, y: 0.0, z: 0.0 },
        KmVec3 { x: 0.0, y: 0.0, z: 0.0 },
        KmVec3 { x: 0.0, y: 0.0, z: 0.0 },
        KmVec3 { x: 0.0, y: 0.0, z: 0.0 },
        KmVec3 { x: 0.0, y: 0.0, z: 0.0 },
        KmVec3 { x: 0.0, y: 0.0, z: 0.0 },
        KmVec3 { x: 0.0, y: 0.0, z: 0.0 },
        KmVec3 { x: 0.0, y: 0.0, z: 0.0 },
    ];

    let mut result: u32 = 2;
    let mut found: bool = false;

    // Fill the corners array
    kmVec3Fill(&mut corners[0], to_check.min.x, to_check.min.y, to_check.min.z);
    kmVec3Fill(&mut corners[1], to_check.max.x, to_check.min.y, to_check.min.z);
    kmVec3Fill(&mut corners[2], to_check.max.x, to_check.max.y, to_check.min.z);
    kmVec3Fill(&mut corners[3], to_check.min.x, to_check.max.y, to_check.min.z);
    kmVec3Fill(&mut corners[4], to_check.min.x, to_check.min.y, to_check.max.z);
    kmVec3Fill(&mut corners[5], to_check.max.x, to_check.min.y, to_check.max.z);
    kmVec3Fill(&mut corners[6], to_check.max.x, to_check.max.y, to_check.max.z);
    kmVec3Fill(&mut corners[7], to_check.min.x, to_check.max.y, to_check.max.z);

    for i in 0..8 {
        if !kmAABB3ContainsPoint(container, &corners[i]) {
            result = 1;
            if found {
                return result;
            }
        } else {
            found = true;
        }
    }

    if !found {
        result = 0;
    }

    result
}
pub fn kmAABB3IntersectsAABB(box_: &KmAabb3, other: &KmAabb3) -> u8 {
    (kmAABB3ContainsAABB(box_, other) != 0) as u8
}
pub fn kmAABB3ExpandToContain<'a>(
    mut pOut: Option<&'a mut KmAabb3>,
    pIn: Option<&'a KmAabb3>,
    other: Option<&'a KmAabb3>,
) -> Option<&'a mut KmAabb3> {
    // Check for None (equivalent to NULL checks in C)
    if pOut.is_none() || pIn.is_none() || other.is_none() {
        return Option::None;
    }

    // Borrow the contents instead of moving them
    let p_out = pOut.as_mut().unwrap();
    let p_in = pIn.unwrap();
    let other = other.unwrap();

    // Modify pOut directly instead of creating a new KmAabb3
    p_out.min.x = p_in.min.x.min(other.min.x);
    p_out.min.y = p_in.min.y.min(other.min.y);
    p_out.min.z = p_in.min.z.min(other.min.z);
    p_out.max.x = p_in.max.x.max(other.max.x);
    p_out.max.y = p_in.max.y.max(other.max.y);
    p_out.max.z = p_in.max.z.max(other.max.z);

    pOut
}
pub fn kmAABB3Initialize<'a>(
    pBox: Option<&'a mut KmAabb3>,
    centre: Option<&'a KmVec3>,
    width: f32,
    height: f32,
    depth: f32,
) -> Option<&'a mut KmAabb3> {
    // Check if pBox is None (equivalent to NULL check in C)
    let pBox = pBox?;

    // Create a default origin point and zero it
    let mut origin = KmVec3 {
        x: 0.0,
        y: 0.0,
        z: 0.0,
    };
    kmVec3Zero(Some(&mut origin));

    // Use the provided centre or fall back to the zeroed origin
    let point = centre.unwrap_or(&origin);

    // Calculate min and max coordinates
    pBox.min.x = point.x - (width / 2.0);
    pBox.min.y = point.y - (height / 2.0);
    pBox.min.z = point.z - (depth / 2.0);
    pBox.max.x = point.x + (width / 2.0);
    pBox.max.y = point.y + (height / 2.0);
    pBox.max.z = point.z + (depth / 2.0);

    Some(pBox)
}
pub fn kmAABB3Centre<'a>(aabb: &KmAabb3, pOut: &'a mut KmVec3) -> &'a mut KmVec3 {
    kmVec3Add(pOut, &aabb.min, &aabb.max);
    let temp = pOut.clone();
    kmVec3Scale(pOut, &temp, 0.5);
    pOut
}
